File: Day3/main.hs
import Data.Char
import Data.Function
import qualified Data.Set as Set
import System.IO

group _ [] = []
group n list = take n list : group n (drop n list)

priority a =
  if isAsciiLower a
    then ord a - ord 'a' + 1
    else ord a - ord 'A' + 27

part1 contents =
  lines contents
    & map
      ( \line ->
          let (first, second) = splitAt (length line `div` 2) line
              a = Set.fromList first
              b = Set.fromList second
              overlap = Set.elemAt 0 $ Set.intersection a b
           in priority overlap
      )
    & sum
    & show

part2 contents =
  lines contents
    & group 3
    & map
      ( \group ->
          let [first, second, third] = map Set.fromList group
              overlapSet = Set.intersection first $ Set.intersection second third
              overlap = Set.elemAt 0 overlapSet
           in priority overlap
      )
    & sum
    & show

main = do
  small_contents <- readFile "in.small"
  contents <- readFile "in"

  putStrLn $ part1 small_contents
  putStrLn $ part1 contents
  putStrLn $ part2 small_contents
  putStrLn $ part2 contents

File: Day3/main.exs
defmodule Day do
  def part1(input) do
    input
    |> Stream.map(fn line ->
      line = String.trim(line)
      {first, second} = String.split_at(line, div(String.length(line), 2))
      first = String.to_charlist(first)
      second = String.to_charlist(second)
      MapSet.to_list(MapSet.intersection(MapSet.new(first), MapSet.new(second)))
    end)
    |> Stream.map(fn [c] -> if c >= ?a && c <= ?z, do: c - ?a + 1, else: c - ?A + 27 end)
    |> Enum.sum()
  end

  def part2(input) do
    input
    |> Stream.map(&String.trim/1)
    |> Stream.chunk_every(3)
    |> Stream.map(fn [a, b, c] ->
      first = String.to_charlist(a)
      second = String.to_charlist(b)
      third = String.to_charlist(c)
      overlap = MapSet.intersection(MapSet.new(first), MapSet.new(second))
      MapSet.to_list(MapSet.intersection(MapSet.new(overlap), MapSet.new(third)))
    end)
    |> Stream.map(fn [c] -> if c >= ?a && c <= ?z, do: c - ?a + 1, else: c - ?A + 27 end)
    |> Enum.sum()
  end
end

small_input = File.stream!("in.small")
input = File.stream!("in")

IO.puts(Day.part1(small_input))
IO.puts(Day.part1(input))

IO.puts(Day.part2(small_input))
IO.puts(Day.part2(input))
